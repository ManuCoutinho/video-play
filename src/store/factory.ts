import { useStore as store } from '.'
import { baseUrl } from '@/constants/baseUrl'

export async function loadInitialData() {
  store.setState({ isLoading: true })
  const controller = new AbortController()
  const signal = controller.signal
  try {
    await fetch(`${baseUrl}/courses/1`, { signal })
      .then((res) => res.json())
      .then((data) =>
        store.setState({
          course: data,
          isLoading: false
        })
      )
  } catch (err) {
    console.error(err)
    //handle error
  } finally {
    controller.abort()
  }
}

export function playNextVideo() {
  const { currentLessonIndex, currentModuleIndex, course } = store.getState()
  const nextLessonIndex = currentLessonIndex + 1
  const nextLesson =
    course?.modules[currentModuleIndex].lessons[nextLessonIndex]

  if (nextLesson) {
    store.setState({
      currentLessonIndex: nextLessonIndex
    })
  } else {
    const nextModuleIndex = currentModuleIndex + 1
    const nextModule = course?.modules[nextModuleIndex]
    if (nextModule) {
      store.setState({
        currentModuleIndex: nextModuleIndex,
        currentLessonIndex: 0
      })
    }
  }
}

export function onPlay(payload: [number, number]) {
  const [moduleIndex, lessonIndex] = payload
  store.setState({
    currentLessonIndex: lessonIndex,
    currentModuleIndex: moduleIndex
  })
}
