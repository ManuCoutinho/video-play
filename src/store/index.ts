import { create } from 'zustand'
import { PlayerStateType } from '@/models/player'
import * as factory from './factory'

type MethodsPlayer = {
  play: (payload: [number, number]) => void
  next: () => void
  load: () => Promise<void>
}
const initialState: PlayerStateType = {
  course: null,
  currentLessonIndex: 0,
  currentModuleIndex: 0,
  isLoading: true
}

const useStore = create<PlayerStateType & MethodsPlayer>(() => {
  return {
    ...initialState,
    play: (payload: [number, number]) => factory.onPlay(payload),
    next: () => factory.playNextVideo(),
    load: () => factory.loadInitialData()
  }
})

export { useStore }
