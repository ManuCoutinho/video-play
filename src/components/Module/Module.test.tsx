import '@testing-library/jest-dom'
import { screen, render, fireEvent } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import { describe, it } from 'vitest'
import { Module } from '.'
import { useStore as store } from '@/store'
import { state } from '@/__mocks__/data/playerState'

const moduleMock = {
  title: 'Iniciando com React',
  moduleIndex: 0,
  amountOfLesson: 2
}
const mockInitialState = store.getState()

describe('<Module/>', () => {
  beforeEach(() => beforeEach(() => store.setState(mockInitialState)))
  it('should render Module correctly ', async () => {
    const { container } = render(<Module {...moduleMock} />)

    act(() => {
      store.setState(state)
    })

    expect(
      screen.getByRole('button', {
        name: /1 iniciando com react 2 aulas/i
      })
    ).toBeInTheDocument()
    expect(screen.getByRole('navigation')).toBeInTheDocument()
    act(() => {
      fireEvent.click(
        screen.getByRole('button', { name: /CSS Modules 13:45/i })
      )
    })
    expect(container).toMatchSnapshot()
  })
})
