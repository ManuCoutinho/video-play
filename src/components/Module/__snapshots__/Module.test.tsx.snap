// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Module/> > should render Module correctly  1`] = `
<div>
  <div
    class="group"
    data-state="open"
  >
    <button
      aria-controls="radix-:r0:"
      aria-expanded="true"
      class="flex w-full items-center gap-4  bg-night p-4"
      data-state="open"
      type="button"
    >
      <div
        class="flex h-10 w-10 rounded-full justify-center items-center  bg-neutral-800 text-sm text-semibold"
      >
        1
      </div>
      <div
        class="flex flex-col gap-1 text-left"
      >
        <strong>
          Iniciando com React
        </strong>
        <span
          class="text-xs text-info flex justify-between"
        >
          2
           aulas
        </span>
      </div>
      <svg
        class="h-5 w-5 ml-auto text-info group-data-[state=open]:rotate-180 transition-transform delay-200 ease-linear hover:text-info/30 transition-color"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M212.24,100.24l-80,80a6,6,0,0,1-8.48,0l-80-80a6,6,0,0,1,8.48-8.48L128,167.51l75.76-75.75a6,6,0,0,1,8.48,8.48Z"
        />
      </svg>
    </button>
    <div
      data-state="open"
      id="radix-:r0:"
      style="transition-duration: 0s; animation-name: none;"
    >
      <nav
        class="relative"
      >
        <button
          class="flex items-center gap-3 text-sm text-info data-[active=true]:text-emerald-300 enabled:hover:text-emerald-200 px-4 py-2"
          data-active="true"
          disabled=""
        >
          <svg
            class="text-emerald-300 w-4 h-4"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M128,26A102,102,0,1,0,230,128,102.12,102.12,0,0,0,128,26Zm0,192a90,90,0,1,1,90-90A90.1,90.1,0,0,1,128,218Zm35.33-95-48-32A6,6,0,0,0,106,96v64a6,6,0,0,0,9.33,5l48-32a6,6,0,0,0,0-10ZM118,148.79V107.21L149.18,128Z"
            />
          </svg>
          <span>
            CSS Modules
          </span>
          <span
            class="ml-auto font-mono text-xs text-secondary"
          >
            13:45
          </span>
        </button>
        <button
          class="flex items-center gap-3 text-sm text-info data-[active=true]:text-emerald-300 enabled:hover:text-emerald-200 px-4 py-2"
          data-active="false"
        >
          <svg
            class="text-primary w-4 h-4"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M250.83,74.71a6,6,0,0,0-6.16.3L206,100.79V72a14,14,0,0,0-14-14H32A14,14,0,0,0,18,72V184a14,14,0,0,0,14,14H192a14,14,0,0,0,14-14V155.21L244.67,181a6,6,0,0,0,9.33-5V80A6,6,0,0,0,250.83,74.71ZM194,184a2,2,0,0,1-2,2H32a2,2,0,0,1-2-2V72a2,2,0,0,1,2-2H192a2,2,0,0,1,2,2Zm48-19.21-36-24V115.21l36-24Z"
            />
          </svg>
          <span>
            Estilização do Post
          </span>
          <span
            class="ml-auto font-mono text-xs text-secondary"
          >
            10:05
          </span>
        </button>
      </nav>
    </div>
  </div>
</div>
`;
